File: scripts/gemma-proxy.js
const express = require('express');
const cors = require('cors');
const axios = require('axios');

const app = express();
const PORT = 3001;

// Enable CORS for all routes
app.use(cors());
app.use(express.json());

// Proxy endpoint for Gemma API
app.post('/api/gemma', async (req, res) => {
  try {
    console.log('📡 Proxying request to Gemma API...');
    
    const gemmaEndpoint = 'https://yoree-gemma-827561407333.europe-west1.run.app/v1beta/models/gemma3:4b:generateContent';
    
    const response = await axios.post(gemmaEndpoint, req.body, {
      timeout: 30000,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'yoree-proxy/1.0'
      }
    });
    
    console.log('✅ Gemma API response received:', response.status);
    res.json(response.data);
    
  } catch (error) {
    console.error('❌ Error proxying to Gemma API:', error.message);
    res.status(500).json({
      error: 'Failed to call Gemma API',
      details: error.message
    });
  }
});

// Health check endpoint
app.get('/api/health', (req, res) => {
  res.json({ status: 'OK', message: 'Gemma proxy is running' });
});

app.listen(PORT, () => {
  console.log(`🚀 Gemma proxy server running on http://localhost:${PORT}`);
  console.log(`📡 Proxy endpoint: http://localhost:${PORT}/api/gemma`);
  console.log(`🏥 Health check: http://localhost:${PORT}/api/health`);
}); 

File: backend/src/gemma_proxy.rs
use axum::{Json, http::StatusCode, response::IntoResponse};
use serde::{Deserialize, Serialize};
use reqwest::Client;

#[derive(Deserialize, Serialize)]
pub struct GemmaRequest {
    pub contents: Vec<Content>,
}

#[derive(Deserialize, Serialize)]
pub struct Content {
    pub parts: Vec<Part>,
}

#[derive(Deserialize, Serialize)]
pub struct Part {
    pub text: String,
}

#[derive(Serialize)]
pub struct GemmaResponse {
    pub candidates: Vec<Candidate>,
}

#[derive(Serialize)]
pub struct Candidate {
    pub index: i32,
    pub content: Content,
    pub finish_reason: String,
}

pub async fn proxy_gemma(request: Json<GemmaRequest>) -> impl IntoResponse {
    let client = Client::new();
    let gemma_url = "https://yoree-gemma-827561407333.europe-west1.run.app/v1beta/models/gemma3:4b:generateContent";
    
    match client
        .post(gemma_url)
        .json(&request.0)
        .header("Content-Type", "application/json")
        .header("User-Agent", "yoree-backend/1.0")
        .send()
        .await
    {
        Ok(response) => {
            match response.json::<serde_json::Value>().await {
                Ok(data) => {
                    (
                        StatusCode::OK,
                        [("Access-Control-Allow-Origin", "*")],
                        Json(data)
                    ).into_response()
                }
                Err(e) => {
                    eprintln!("Error parsing Gemma response: {}", e);
                    (
                        StatusCode::INTERNAL_SERVER_ERROR,
                        Json(serde_json::json!({
                            "error": "Failed to parse Gemma response"
                        }))
                    ).into_response()
                }
            }
        }
        Err(e) => {
            eprintln!("Error calling Gemma API: {}", e);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(serde_json::json!({
                    "error": "Failed to call Gemma API"
                }))
            ).into_response()
        }
    }
}

pub async fn handle_options() -> impl IntoResponse {
    (
        StatusCode::OK,
        [
            ("Access-Control-Allow-Origin", "*"),
            ("Access-Control-Allow-Methods", "POST, OPTIONS"),
            ("Access-Control-Allow-Headers", "Content-Type"),
        ]
    ).into_response()
} 